carapace-server/src/main/java/org/carapaceproxy/server/certificates/DynamicCertificatesManager.java

```diff
@@ -362,9 +376,108 @@ private void certificatesLifecycle() {
+        if (challenge == null) {
+            cert.setState(VERIFIED);
+        } else {
+            LOG.log(Level.INFO, "Pending challenge data for domain {0}: {1}", new Object[]{cert.getDomain(), challenge.getJSON()});
+            if (cert.isWildcard()) {
+                dnsChallegesCreationStatus.put(cert.getDomain(), new DnsChallengeCreationStatus());
+                Dns01Challenge ch = (Dns01Challenge) challenge;
+                r53Client.createDnsChallengeForDomain(cert.getDomain(), ch.getDigest(), (res, err) -> {
+                    if (err == null) {
+                        LOG.log(Level.INFO, "Created new TXT DNS challenge-record for domain {0}. Details: {1}", new Object[]{cert.getDomain(), res});
+                    } else {
+                        DnsChallengeCreationStatus status = dnsChallegesCreationStatus.get(cert.getDomain());
+                        if (status != null) {
+                            status.setError();
+                        }
+                        LOG.log(Level.INFO, "Creation of TXT DNS challenge-record for domain {0} FAILED. Reason: {1}", new Object[]{cert.getDomain(), err});
+                    }
+                });
+                cert.setState(DNS_CHALLENGE_WAIT);
+            } else {
+                Http01Challenge ch = (Http01Challenge) challenge;
+                store.saveAcmeChallengeToken(ch.getToken(), ch.getAuthorization()); // used for incoming acme-challenge requests
+                challenge.trigger();
+                cert.setState(VERIFYING);
+            }
+            cert.setPendingChallengeData(challenge.getJSON());
+        }
+    }
+
+    private Challenge getChallengeFromCertificate(CertificateData cert) throws AcmeException {
+        try {
+            JSON challengeData = JSON.parse(cert.getPendingChallengeData());
+            LOG.log(Level.FINE, "CHALLENGE: {0}.", challengeData);
+            return cert.isWildcard()
+                    ? new Dns01Challenge(acmeClient.getLogin(), challengeData)
+                    : new Http01Challenge(acmeClient.getLogin(), challengeData);
+        } catch (IOException | AcmeException e) {
+            return null;
+        }
+    }
+
+    private void checkDNSChallengeAvailabilityForCertificate(Dns01Challenge challenge, CertificateData cert, DnsChallengeCreationStatus creationStatus) throws AcmeException {
+        String domain = cert.getDomain();
+        if (creationStatus.isCompleted()) {
+            cert.setState(VERIFYING);
+            dnsChallegesCreationStatus.remove(domain);
+            challenge.trigger();
+        } else if (creationStatus.testReachability()) {
+            cert.setState(REQUEST_FAILED);
+            dnsChallegesCreationStatus.remove(domain);
+            cleanupChallengeForCertificate(challenge, cert);
+        } else {
+            r53Client.isDnsChallengeForDomainAvailable(domain, challenge.getDigest(), (available, err) -> {
```

Comment:
you should block and wait for this function to complete.
otherwise you will end in concurrent handling of the DnsChallengeCreationStatus data structure